src/merkle/store/tests.rs

```diff
@@ -647,6 +647,149 @@ fn test_constructors() -> Result<(), MerkleError> {
+        assert_eq!(d, store.get_leaf_depth(root, k).unwrap());
+
+        // insert and assert the correct depth
+        root = store.set_node(root, index, node).unwrap().root;
+        assert_eq!(64, store.get_leaf_depth(root, k).unwrap());
+    }
+}
+
+#[test]
+fn get_leaf_depth_works_with_incremental_depth() {
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // insert some path to the left of the root and assert it
+    let key = 0b01001011_10110110_00001101_01110100_00111011_10101101_00000100_01000001_u64;
+    assert_eq!(0, store.get_leaf_depth(root, key).unwrap());
+    let depth = 64;
+    let index = NodeIndex::new(depth, key).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // flip the key to the right of the root and insert some content on depth 16
+    let key = 0b11001011_10110110_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(1, store.get_leaf_depth(root, key).unwrap());
+    let depth = 16;
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // attempt the sibling of the previous leaf
+    let key = 0b11001011_10110111_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(16, store.get_leaf_depth(root, key).unwrap());
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+
+    // move down to the next depth and assert correct behavior
+    let key = 0b11001011_10110100_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    assert_eq!(15, store.get_leaf_depth(root, key).unwrap());
+    let depth = 17;
+    let index = NodeIndex::new(depth, key >> (64 - depth)).unwrap();
+    let node = [Felt::new(key); WORD_SIZE];
+    root = store.set_node(root, index, node).unwrap().root;
+    assert_eq!(depth, store.get_leaf_depth(root, key).unwrap());
+}
+
+#[test]
+fn get_leaf_depth_works_with_small_depth() {
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+    let node = [ONE; WORD_SIZE];
+
+    // insert some random, 8 depth keys. `a` diverges from the first bit
+    let a = 0b01101001_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let b = 0b10011001_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let c = 0b10010110_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
+    let d = 0b11110110_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64;
```

Comment:
Why not declare these without trailing zeros? i.e.,
```Rust
let a = 0b0110_1001_u64;
```

Reply:
That one would be in fact a different path:

```rust
let a = 0b00000000_00000000_00000000_00000000_00000000_00000000_00000000_01101001_u64;
```

The most significant bit (i.e. leftmost) is the first lint between the root and the first depth. If we declare without the right zeroes, we will always have a root starting at zero (hence, depth will always range from 57 to 64)

Using this leftmost declaration is also an unmistakable documentation on how we treat the bits (i.e. we use the most significant ones for the root, and least significant ones for the leaves)

Reply:
The way I think about this test case is that we are dealing with trees of small depth (up to 8), and thus, the keys should reflect that. Value `0b01101001_00000000_00000000_00000000_00000000_00000000_00000000_00000000_u64` is not a valid index for a depth 8 tree and we need to adjust it (i.e., `a >> 56`) to make it valid.

So, basically, we start with an invalid value and then do an extra operation to make it valid. I'd prefer to start out with a valid value. And again, I think the key difference here is that I'm thinking of this test as working with depth 8 trees only.

Reply:
We have to keep in mind that this `get_leaf_path` will work only for trees of depth `64`, as we treat the empty sub-tree constants that way too

When we say `depth 8`, we actually mean a branch to the root of depth 64 that was inserted into the storage, that happens to contain only 8 nodes.

To double check this is the desired effect, the test would look like this:

```rust
// insert some random, 8 depth keys. `a` diverges from the first bit
let a = 0b01101001_u64;
let b = 0b10011001_u64;
let c = 0b10010110_u64;
let d = 0b11110110_u64;

for k in [a, b, c, d] {
    let index = NodeIndex::new(8, k).unwrap();
    root = store.set_node(root, index, node).unwrap().root;
}

// assert all leaves returns the inserted depth
for k in [a, b, c, d] {
    assert_eq!(8, store.get_leaf_depth(root, k << 56).unwrap());
}
```

We need to do that `k << 56` because that argument expects a full path of `64` bits

Reply:
Double checked. We need to do that because the go "left" for the first 56 bits, and then we start the intended path.

The path from the root starts at the most significant bit of the number

Reply:
To avoid this ambiguity, we could name the method `get_leaf_depth_tree_64`, or `get_vacant_64_depth` as @hackaugusto suggested

Reply:
Not at all! The only downside is we can't traverse more than `64` bits, so we return an error if the user provide anything beyond 64. The rest is just replacing `MAX_DEPTH` constant by this parameter. Lemme push the change so we can visualize it

Reply:
One note: should we call the parameter `tree_depth` instead of just `depth`?

Reply:
I called `max_depth`, but `tree_depth` fits better